<a href="https://colab.research.google.com/github/Jialan8020/Assign8/blob/master/question4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Fit a multilayer neural network model on the SHHS data to predict whether rdi4p > 7 using the variables from the previous HW assigment. Compare the test set accuracy with the penalized model from your previous HW.

In [0]:
import pandas as pd
import tensorflow as tf
import numpy as np
from sklearn.model_selection import cross_val_score
from tensorflow import keras
from sklearn.model_selection import train_test_split

In [8]:
dat = pd.read_csv("shhs.txt",sep="\t")
dat = dat.dropna()
dat.head(4)

pptid  waist  COPD15  ASTHMA15  ...  age_s1  smokstat_s1   WASO     bmi_s1
3       4   85.0     0.0       0.0  ...      48          0.0   43.5  20.185185
4       5   76.0     0.0       0.0  ...      66          2.0  100.5  23.309053
8       9   99.0     0.0       0.0  ...      69          2.0   52.5  25.817447
14     15   80.0     0.0       0.0  ...      58          2.0  132.5  23.937672

[4 rows x 30 columns]

In [9]:
# Extract the rdi4p column
rdi = pd.DataFrame(dat["rdi4p"])
rdi

rdi4p
3      0.797342
4      2.756757
8     11.948791
14    16.254980
15     3.102041
...         ...
5795   2.108346
5798   0.469667
5799  34.415781
5801   2.163462
5802   1.038319

[2897 rows x 1 columns]

In [10]:
rdi[rdi < 7] = 0
rdi[rdi > 7] = 1
rdi = rdi.astype(int)
rdi


rdi4p
3         0
4         0
8         1
14        1
15        0
...     ...
5795      0
5798      0
5799      1
5801      0
5802      0

[2897 rows x 1 columns]

In [18]:
X = dat.drop(["rdi4p","pptid"], axis=1)
X

waist  COPD15  ASTHMA15  slp_lat  ...  age_s1  smokstat_s1   WASO     bmi_s1
3      85.0     0.0       0.0     14.0  ...      48          0.0   43.5  20.185185
4      76.0     0.0       0.0      6.5  ...      66          2.0  100.5  23.309053
8      99.0     0.0       0.0     20.0  ...      69          2.0   52.5  25.817447
14     80.0     0.0       0.0     66.5  ...      58          2.0  132.5  23.937672
15    100.0     0.0       0.0     69.5  ...      60          0.0   84.0  29.475309
...     ...     ...       ...      ...  ...     ...          ...    ...        ...
5795  117.0     0.0       0.0     29.5  ...      71          0.0   55.5  35.790598
5798  103.0     0.0       0.0     10.5  ...      54          2.0   13.0  32.414213
5799   99.0     0.0       0.0     11.5  ...      66          0.0   98.5  25.693134
5801   91.0     0.0       0.0      4.0  ...      55          0.0   30.0  24.228571
5802   93.0     0.0       0.0      8.0  ...      53          0.0   76.0  25.374483

[2897 rows x 28 columns]

In [0]:
# Separate training and test data
X_train, X_test, y_train, y_test = train_test_split(X, rdi, test_size=0.2)

In [0]:
model = tf.keras.Sequential([
  tf.keras.layers.Dense(64, activation='relu',input_dim=28),
  tf.keras.layers.Dense(10, activation='relu'),
  tf.keras.layers.Dense(1,activation='sigmoid')
])
model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=['mse'])

In [22]:
model.fit(X_train, y_train, epochs=5)

Epoch 1/5
73/73 [==============================] - 0s 1ms/step - loss: 6.7350 - mse: 0.4088
Epoch 2/5
73/73 [==============================] - 0s 1ms/step - loss: 1.1117 - mse: 0.3060
Epoch 3/5
73/73 [==============================] - 0s 1ms/step - loss: 0.8616 - mse: 0.2674
Epoch 4/5
73/73 [==============================] - 0s 1ms/step - loss: 0.9494 - mse: 0.2821
Epoch 5/5
73/73 [==============================] - 0s 1ms/step - loss: 0.7299 - mse: 0.2360


In [23]:
model.evaluate(X_test,y_test,verbose=2)

19/19 - 0s - loss: 0.8608 - mse: 0.2992


[0.8607730865478516, 0.2991858124732971]

**Comparison with HW 7**

The best cross validated error of Lasso regression is 0.8683.

The best cross validated error of Ridge regression is 0.8694.

The best cross validated error of elastic net is 0.8687.

The best cross validated error of no penalty is 0.8691.

However, by using neural network algorithm, we achieved a mse of 0.2992, which is significantly smaller than the logistic regression model.